### DISCUSSION TITLE
Is it possible to listen to what has exactly changed in the DB?

### DISCUSSION THREAD

**Discussion by miguelcmedeiros** [simolus3/drift#1258] [Q&A]
I was wondering if there's a way for an app to attach a listener to the DB and get events with information about what has changed? 

This would include for instance, which table was updated, the kind of update (update, delete, insert) and what was changed.

**Comment by simolus3** [comment]
You can use [`tableUpdates`](https://pub.dev/documentation/moor/latest/moor/DatabaseConnectionUser/tableUpdates.html) to obtain a stream that emits a null item for a fixed filter.
I think we could change that stream to return the specific table update, but it's not clear to me how advanced features like update propagation would work. For instance, if you declare a table with a foreign key and  `ON DELETE CASCADE`, a delete on one table may cascade to a delete on another table. Should we emit two items in that case? The current stream would just emit a single item for that change.

**Comment by long1eu** [comment]
This is also I question that I have. Do we get delete updates when using watch?

**Reply by long1eu** [reply]
https://www.sqlite.org/c3ref/update_hook.html

**Reply by simolus3** [reply]
> Do we get delete updates when using watch?

Yes, you should get them. Moor does not use update hooks because they don't work on all tables, it has its own implementation.

**Comment by long1eu** [comment]
Awesome. Stupid question, how do you know something was removed from the
watch stream?

**Reply by simolus3** [reply]
Moor uses a fairly conservative algorithm for stream queries which mostly involves saying "this stream reads from tables x, y and z, so it should update when any of those tables is written to". 
For instance, the `DeleteStatement` class in moor will inform the database that there [was a delete](https://github.com/simolus3/moor/blob/ac30bb598946b0fbfb9e278e73cf9451c01e237c/moor/lib/src/runtime/query_builder/statements/delete.dart#L44-L45) on table x, which will get propagated to the stream query. This may sound inefficient, but it's good enough and more accurate implementations would be pretty complicated (which also increases code size etc.). Moor is a bit more sophisticated because we need to support views, foreign key constraints with `ON DELETE CASCADE` and triggers which have an impact on how stream queries are propagated.

**Comment by long1eu** [comment]
I absolutely understand that. But my issue is that I don't understand how I
differenciate between event's in the stream. All I get is a map with the
values from.the row, how do I know if this is a delete, update or insert?

**Reply by simolus3** [reply]
In 8c134138d2db24e30661be1e0e0d8fcea7b419cf, I've changed the stream to return a set of relevant table updates so that you can get the detailed information.

**Reply by long1eu** [reply]
Still this looks a bit different from what I was expecting. For example with firestore you have a Document, in this case is a row, and you have a DocumentChange, witch contains the Document(row) and the change type(add, update, delete). All of this come from the same listen(watch) stream.

**Reply by simolus3** [reply]
Sorry, but I don't think we can reasonably provide this feature in moor. Implementing it in Dart is impossible (think of triggers and other factors that make it impossible for us to determine the right id in all scenarios), and native update hooks are not available on all platforms.

You could run a regular query stream and then diff results to obtain changes if that's what you need.

**Comment by long1eu** [comment]
I'm already doing something very similar but more rudimentary by just
keeping track of the hashCodes that were emited, so if the item was already
emited but the hash changed, then it's an update otherwise it's an insert.
The tricky part is with deletes, since I haven't find a way to know what
I'd had been removed.

**Comment by long1eu** [comment]
Thanks for all the awesome work you are doing.
